src/hooks/useForm.js

```diff
@@ -0,0 +1,218 @@
+    if (!isValid) {
+      setState({
+        ...state,
+        errors: {
+          ...validationErrors
+        }
+      });
+      return;
+    }
+
+    try {
+      await submitForm();
+    } catch (error) {
+      // Note that the state will NOT be set if the  error is
+      // caught inside the callback (submitForm) and not re-thrown
+      setState({
+        ...state,
+        submitError: error.message
+      });
+    }
+  }
+
+  /**
+   * Reset the form to its initial value.
+   * @returns {void} Nothing.
+   */
+  function handleReset() {
+    setState(init(schema, initialFormState));
+  }
+
+  /**
+   * Set the new value onchange, and validate property or matching properties.
+   * @param {Event} event - The onChange event.
+   * @returns {void} Nothing.
+   */
+  function handleInputChange(event) {
+    const { form } = state;
+    const { value, name } = event.target;
+
+    // Ah, the good old days!
+    setState({
+      ...state,
+      form: {
+        ...form,
+        [name]: value
+      },
+      errors: {
+        ...validateProperty(name, value)
+      }
+    });
+  }
+
+  /**
+   * Validate one or pair of corresponding properties.
+   * @param {string} name - The property name.
+   * @param {string} value - The property value.
+   * @returns {object} All the errors in the entire form.
+   */
+  function validateProperty(name, value) {
+    const { errors, form } = state;
+    let allErrors = {
+      ...errors
+    };
+    const matchingProperty = getMatchingProperty(name, schema);
+
+    // No matching property, just validate this one property
+    if (!matchingProperty) {
+      const { isValid, errors: propertyErrors } = validate(value, schema[name]);
+
+      isValid
+        ? delete allErrors[name]
+        : (allErrors = {
+            ...allErrors,
+            [name]: propertyErrors
+          });
+
+      return { ...allErrors };
+    }
+
+    // Matching properties present. ex: password & confirm password
+    const matchingValues = {
+      [name]: value,
+      [matchingProperty]: form[matchingProperty]
+    };
+
+    const matchingValuesSchema = {
+      [name]: schema[name],
+      [matchingProperty]: schema[matchingProperty]
+    };
+
+    // Clear previous errors on matching properties before
+    // potentially re-setting them
+    delete allErrors[name];
+    delete allErrors[matchingProperty];
+
+    const { errors: propertyErrors } = validate(matchingValues, matchingValuesSchema);
+
+    return {
+      ...allErrors,
+      ...propertyErrors
+    };
+  }
+
+  return {
+    ...state,
+    handleReset,
+    handleSubmit,
+    handleInputChange
+  };
+}
+
+/**
+ * Derive state from the given schema.
+ * @param {object} schema - The given schema.
+ * @param {object} initialFormState - The initial values of the form properties.
+ * @returns {...UseFormState} The useForm initial state.
+ */
+function init(schema, initialFormState) {
```

Comment:
i feel like init should be at the top since it's kinda a constructor and that's where i initially looked when i saw 'init'

Reply:
Fair point. I can move that up.